docs/explanation/locks.rst

```diff
@@ -0,0 +1,236 @@
+Essentially, RT-mutexes can be used to implement spinlock_t and read-write locks, providing the added advantage of priority inheritance where applicable.
+
+semaphore
+~~~~~~~~~~
+
+The ``semaphore`` (defined in `include/linux/semaphore.h`_) structure discussed here follow the counting semaphore structure (as the binary semaphore equivalent is essentially a mutex).
+It's a foundational implementation intended for use within other kernel lock structures.
+Internally, it relies on a :ref:`raw_spinlock` and a counter.
+Their primary use cases involve serialization and waiting.
+It’s important to note that newer mechanisms, such as mutexes for serialization and `completions`_ for waiting, are generally preferable to traditional semaphores.
+
+
+rw_semaphore
+~~~~~~~~~~~~
+
+``rw_semaphore`` (Read-Write Semaphore, defined on `include/linux/rwsem.h`_) is a variant of the read/write lock for sleeping locks.
+The goal of this locking mechanism is to reduce the overhead caused by acquiring and releasing locks by minimizing the amount of locking required.
+A read-write semaphore can be held by many readers but only one writer.
+This significantly reduces the number of locks needed for various types of data.
+
+ww_mutex
+~~~~~~~~
+
+``ww_mutex`` (Wound/Wait Mutex found on `include/linux/ww_mutex.h`_) are designed to handle complex synchronization scenarios, especially in GPU drivers where multiple buffers are shared across contexts, processes, and devices.
+Traditional mutexes can lead to deadlocks due to unpredictable buffer access orders.
+WW-Mutexes prevent deadlocks using a reservation ticket system, where the oldest task wins in case of contention.
+Two algorithms, Wound-Wait and Wait-Die, manage conflicts, with Wound-Wait generally reducing rollbacks.
+WW-Mutexes support flexible locking strategies, making them suitable for dynamic and unordered lock acquisition scenarios.
+
+
+percpu_rw_semaphore
+~~~~~~~~~~~~~~~~~~~
+
+The ``percpu_rw_semaphore`` (defined in `include/linux/percpu-rwsem.h`_) is a modern read-write semaphore designed with optimization for read-heavy workloads.
+Traditional read-write semaphores suffer from performance degradation when multiple cores acquire the lock for reading, as the cache line containing the semaphore bounces between the L1 caches of those cores.
+With percpu_rw_semaphore, reading is extremely fast due to the use of RCU (`Read-Copy-Update`_), which eliminates the need for atomic instructions in the lock and unlock paths.
+However, locking for writing is significantly more expensive, as it involves calling synchronize_rcu(), which can take hundreds of milliseconds.
+
+
+CPU Local Locks
+---------------
+
+On non-PREEMPT_RT kernels, ``local_lock`` functions are wrappers around preemption and interrupt-disabling primitives.
+Unlike other locking mechanisms, disabling preemption or interrupts is strictly a CPU-local concurrency control technique and is not suitable for managing inter-CPU concurrency.
+
+local_lock_t
```

Comment:
Why this differs in the kernel docs? There it is referred to as `local_lock`.

Reply:
The first mentions the `functions` the second mentions the lock type `local_lock_t`.